CorsixTH/Lua/rooms/training.lua

```diff
@@ -81,16 +81,29 @@ function TrainingRoom:calculateTrainingFactor(objects)
     end
     return counts
   end
-  -- Work out total training factor (averaged of all relevant items)
-  -- param factors(table) An array containing a count of each object and its raw training value
+  -- Work out total training factor using two averaging methods, choosing the best one
+  -- param factors(table) An array containing a count of each object and its raw
+  -- training value
   -- returns the total effect
   local function calculateTotalEffect(factors)
-    local count, total = 0, 0
-    for i = 1, #factors do
-      count = count + factors[i].count
-      total = total + (factors[i].count * factors[i].value)
+    assert(factors,
+        "Unable to determine the training factor of a training room! " ..
+        "Reason: No inputs given")
+    local total_objects = 0
+    local total_value = 0
+    local num_factors = #factors
+    for i = 1, num_factors do
+      total_objects = total_objects + factors[i].count
+      -- Check we have at least one of this object and action accordingly
+      local multiplier = factors[i].count > 0 and (math.log(factors[i].count) + 1) or 0
+      total_value = total_value + (multiplier * factors[i].value)
     end
```

Comment:
In this entire loop you only use `i` for indexing into `factors`. Is there any particular reason you cannot use `ipairs()` ?

Reply:
I think I was having a moment when writing it, so I'll change it over.